### ISSUE TITLE
Unmountable filesystem after upgrading, fsck fails [8277b87958c5]

### ISSUE THREAD

**Issue by antihutka** [koverstreet/bcachefs#428]
I was running on 5.10 (e455e478dd8c) until yesterday and tried to upgrade. I couldn't get 5.18 running because of a Nvidia compatibility issue, so I tried to at least to upgrade to the 5.17 branch (7a3837e719b6).
The filesystem failed to mount, I built the best matching bcachefs-tools I could find (962390c0b2e9) and ran fsck, which also failed.
I also tried to build the current version of bcachefs-tools (b4afc1fa2667), but even with that fsck seems to fail in the same way.
Finally, I tried to build the latest kernel (8277b87958c5), boot it in a VM and tried to mount. This failed too.
Since I hit a build error in 5.18 caused by stack_trace_save_tsk, I was forced to comment out a couple of lines in fs/bcachefs/debug.c, so there might be some missing output in the kernel log - I'm not sure.

All of the logs are here: https://pastebin.com/J10La6TB
Journal and metadata dumps:
https://hammy.sk/download/journal.log.zst
https://hammy.sk/download/metadata.dump.0.qcow2.xz
https://hammy.sk/download/metadata.dump.1.qcow2.xz

**Comment by antihutka** [comment]
Mounting with -o nochanges,norecovery,ro fails in a different way:
```[   38.314470] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): recovering from unclean shutdown
[   45.969422] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): ja->sectors_free == ca->mi.bucket_size
[   45.969585] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): cur_idx 322/1024
[   45.969729] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): bucket_seq[321] = 3150455
[   45.969845] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): bucket_seq[322] = 3150459
[   45.969933] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): bucket_seq[323] = 3127427
[   46.743564] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): journal read done, 1557610 keys in 415 entries, seq 9042174
[   52.146199] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): mounted version=inode_backpointers opts=ro,compression=lz4,foreground_target=ssd,background_target=remote,promote_target=ssd,noinodes_32bit,noshard_inode_numbers,noinodes_use_key_cache,journal_reclaim_delay=1000,nochanges,norecovery
[   52.146292] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): missing subvolume 1
[   52.146451] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): inconsistency detected - emergency read only
[   52.146547] bcachefs (afdf8c2e-f853-45ec-872d-7264d9549268): error mounting: error getting root inode -2
```

**Comment by jpsollie** [comment]
what does "bcachefs set-option --version_upgrade (devices)" do?

**Comment by jpsollie** [comment]
you could do a bcachefs fsck -v -f (devices) to see whether the upgrading worked correctly
if that's ok, your filesystem should work again

**Comment by antihutka** [comment]
Unfortunately this failed again: https://pastebin.com/L42yhQtT

**Comment by jpsollie** [comment]
if you mount with -o ro,very_degraded, is your data readable?

**Comment by antihutka** [comment]
No, mounting still BUGs in a similar way as before: https://pastebin.com/MJZVF5jJ

**Comment by jpsollie** [comment]
@koverstreet, @YellowOnion  any ideas?

**Comment by antihutka** [comment]
I managed to write a tool and recovered all of the data on the filesystem, and as far as I can tell all of it is intact. I can recreate the filesystem and restore all of the data, unless you want me to keep it around to try other things.
I had to patch a few snapshot-related things in the code to stop `bch2_read` from crashing, but other than that I didn't hit any issues. Compared to a newly created filesystem, the filesystem doesn't have any subvolume and snapshot keys. Could this mean the initial filesystem upgrade failed for some reason?

**Comment by koverstreet** [comment]
ancient bug, should be fixed
